asab/sentry/service.py

```diff
@@ -0,0 +1,265 @@
+import asab
+import sys
+import os
+import re
+import logging
+import json
+
+L = logging.getLogger(__name__)
+
+
+try:
+	import sentry_sdk
+	import sentry_sdk.integrations.aiohttp
+	import sentry_sdk.integrations.asyncio
+	import sentry_sdk.integrations.logging
+except ModuleNotFoundError:
+	L.critical("Package for Sentry SDK not found. Install it with: pip install sentry-sdk")
+	sys.exit(1)
+
+
+class SentryService(asab.Service):
+	"""
+	Service for Sentry SDK integration.
+
+	Sentry is an error tracking and performance monitoring platform.
+	When the service is initialized and `data_source_name` is set,
+	all unhandled exceptions, error log messages are sent as Events to sentry.io,
+	together with lines of code where the error happened, structured data and values of variables.
+
+	Configuration:
+	```ini
+	[sentry]
+	data_source_name=
+	```
+
+	Examples:
+	```python
+	sentry_svc = app.get_service("SentryService")
+
+	try:
+		call_collapsing_function()
+	except Exception as e:
+		sentry_svc.capture_exception(e)
+
+	sentry_svc.capture_message("User was deleted from database.")
+	```
+
+	"""
+
+	def __init__(self, app, service_name: str):
+		super().__init__(app, service_name)
+
+		# DATA SOURCE NAME (DSN)
+		# format: https://<public key>@o<secret key>.ingest.sentry.io/<project id>
+		# DSN is automatically generated when new project is created
+		# and can be modified: Settings > Client Keys (DSN) > Key Details
+		self.DataSourceName = asab.Config.get("sentry", "data_source_name", fallback="")
+		if len(self.DataSourceName) == 0:
+			self.DataSourceName = os.getenv("SENTRY_DSN", "")
+		if len(self.DataSourceName) == 0:
+			L.error("Data source name is not set. Specify it via SENTRY_DSN env variable or in configuration: [sentry] data_source_name.")
+		# LOGGING LEVELS
+		levels = {
+			"debug": logging.DEBUG,
+			"info": logging.INFO,
+			"notice": asab.LOG_NOTICE,
+			"warning": logging.WARNING,
+			"error": logging.ERROR,
+			"critical": logging.CRITICAL
+		}
+
+		self.LoggingBreadCrumbsLevel = levels.get(asab.Config.get("sentry:logging", "breadcrumbs").lower(), "notice")
+		self.LoggingEventsLevel = levels.get(asab.Config.get("sentry:logging", "events").lower(), "error")
+
+		# ENVIRONMENT (e.g. "production", "testing", ...)
+		self.Environment = asab.Config.get("sentry", "environment")  # default: "develop"
+
+		# RELEASE
+		# Release can be obtained from MANIFEST.json if exists
+		path = asab.Config.get("general", "manifest")
+		if path == "":
+			if os.path.isfile("/app/MANIFEST.json"):
+				path = "/app/MANIFEST.json"
+			elif os.path.isfile("/MANIFEST.json"):
+				path = "/MANIFEST.json"
+			elif os.path.isfile("MANIFEST.json"):
+				path = "MANIFEST.json"
+
+		if len(path) != 0:
+			try:
+				with open(path) as f:
+					manifest = json.load(f)
+			except Exception as e:
+				L.exception("Error when reading manifest for reason {}".format(e))
```

Comment:
Unterminated exception handling ... you should set `manifest = None` before main `if` here, so that his is defensible.